pcicrawler/cli.py

```diff
@@ -65,6 +74,283 @@ def jsonify(dev, hexify=False, vpd=False, aer=False):
+        
+        def run(self):
+            return super().run(self._device.express_link.capable_speed)
+
+
+    class check_capable_link_width(OCPTestStep):
+        def __init__(self,
+                     device: PCIDevice,
+                     expected_results: str,
+                     ocp_run: tv.TestRun):
+            self._test_name = "pci_capable_link_width_check"
+            super().__init__(device, expected_results, ocp_run)
+        
+        def run(self):
+            return super().run(self._device.express_link.capable_width)
+
+
+    class check_address(OCPTestStep):
+        def __init__(self,
+                     device: PCIDevice,
+                     expected_results: str,
+                     ocp_run: tv.TestRun):
+            self._test_name = "pci_address_check"
+            super().__init__(device, expected_results, ocp_run)
+
+        def run(self):
+            dut = tv.Dut(id=str(self._device), name="PCIe Device")
+            with self._ocp_run.scope(dut=dut):
+                step = self._ocp_run.add_step(self._test_name)
+                if str(self._device) not in str(self._expected_result):
+                    step.add_diagnosis(tv.DiagnosisType.FAIL, 
+                                    verdict="Test {} did not find {} in expected list {}.".format(
+                                        self._test_name,
+                                        self._device,
+                                        self._expected_result))
+                else:
+                    step.add_diagnosis(tv.DiagnosisType.PASS, 
+                                    verdict="Test {} for device {} PASSED".format(
+                                        self._test_name,
+                                        self._device))
+
+
+    class check_AER(OCPTestStep):
+        def __init__(self,
+                     device: PCIDevice,
+                     expected_results: str,
+                     ocp_run: tv.TestRun):
+            self._test_name = "pci_AER_check"
+            super().__init__(device, expected_results, ocp_run)
+
+        def run(self):
+            return super().run(self._device.express_aer)
+
+
+    class TestSet(object):
+        def __init__(self, duts: List[List[PCIDevice]], validate: Dict):
+            self.duts = duts
+            self.validators = validate
+
+        def __str__(self):
+            return str(self.duts)
+
+
+    def run(self):
+        duts = []
+        # Process input json
+        with open(self._json_file, 'r') as fh:
+            try:
+                input_json = json.load(fh)
+                duts = input_json['duts']
+            except Exception as _:
+                raise("ERROR: Cannot parse file {}".format(self._json_file))
+        filtered_duts = []
+        for dut in duts:
+            identifier = dut['identifiers']
+            potential_duts = list(self._devs)
+            # Go look for this device and get the data
+            if 'address' in identifier:
```

Comment:
im kinda curious why this filtering exists here; couldnt the input data be already filtered? it looks to me like youre taking input data and then use some aspect that also exists in that data to filter it, which points me to the question

Reply:
this is to filter the input data, I'm taking the existing devs list and matching it to the requirements in the json, based on the discussed JSON with the team, technically a device can exist in multiple test sets, it was a little oversight on my part when i defined the input json but it doesn't matter